core/src/main/java/tech/beshu/ror/requestcontext/SerializationTool.java

```diff
@@ -45,70 +50,94 @@
+    try {
+      Class clazz = Class.forName(configuredSerializer.get());
+      Constructor constr = clazz.getConstructor(new Class[0]);
+      constr.setAccessible(true);
+      AuditLogSerializer serializerInstance = (AuditLogSerializer) constr.newInstance(new Object[0]);
+      this.customSerializer = Optional.of(serializerInstance);
+      logger.info("Using custom serializer: " + serializerInstance.getClass().getName());
+      return;
+    } catch (InstantiationException | IllegalAccessException | ClassNotFoundException | NoSuchMethodException | InvocationTargetException e) {
+      logger.error("Error picking the custom serializer, proceeding with default.", e);
+      this.customSerializer = Optional.empty();
+      return;
+    }
+  }
 
-    map.put("req_method", req.getMethod());
-    map.put("headers", req.getHeaders().keySet());
-    map.put("path", req.getUri());
+  public String mkIndexName() {
+    return "readonlyrest_audit-" + indexNameFormatter.format(Calendar.getInstance().getTime());
+  }
 
-    map.put("user", req.getLoggedInUser().isPresent() ? req.getLoggedInUser().get().getId() : null);
+  public String toJson(ResponseContext rc) {
+    RequestContextShim req = rc.getRequestContext();
+
+    AuditLogContext logContext = new AuditLogContext()
+      .withId(req.getId())
+      .withAction(req.getAction())
+      .withPath(req.getUri())
+      .withHeaders(req.getHeaders().keySet())
+      .withAclHistory(req.getHistoryString())
+      .withContentLen(req.getContentLength())
+      .withContentLenKb(req.getContentLength() == 0 ? 0 : req.getContentLength() / 1024)
+      .withOrigin(req.getRemoteAddress())
+      .withErrorType(rc.getError() != null ? rc.getError().getClass().getSimpleName() : null)
+      .withErrorMessage(rc.getError() != null ? rc.getError().getMessage() : null)
+      .withType(req.getType())
+      .withTaskId(Math.toIntExact(req.getTaskId()))
+      .withReqMethod(req.getMethodString())
+      .withUser(req.getLoggedInUserName().isPresent() ? req.getLoggedInUserName().get() : null)
+      .withIndices(req.involvesIndices() ? req.getIndices() : Collections.emptySet())
+      .withTimestamp(zuluFormat.format(rc.getRequestContext().getTimestamp()))
+      .withProcessingMillis(rc.getDurationMillis())
+      .withMatchedBlock(rc.getReason())
+      .withFinalState(rc.finalState().name());
 
-    map.put("action", req.getAction());
-    map.put("indices", req.involvesIndices() ? req.getIndices() : Collections.emptySet());
-    map.put("acl_history", req.getHistory().toString());
 
-    // The cumbersome, awkward security handling in Java. (TM) #securityAgainstProductivity
     final String[] res = new String[1];
 
     AccessController.doPrivileged((PrivilegedAction<Void>) () -> {
       try {
-        res[0] = mapper.writeValueAsString(map);
+        if (customSerializer.isPresent()) {
+          res[0] = mapper.writeValueAsString(customSerializer.get().createLoggableEntry(logContext));
```

Comment:
By passing the logContext, you're limiting what we can add to our audit log entry by the fields that are included in logContext. Eg we would not be able to include the value of whitelisted headers (which is what we need to have).
createLoggableEntry should accept ResponseContext as a parameter, not AuditLogContext.